src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.NativeAot.cs

```diff
@@ -210,6 +210,27 @@ public static void SetLastPInvokeError(int error)
             PInvokeMarshal.t_lastError = error;
         }
 
+        /// <summary>
+        /// Begins tracking system call errors on the calling thread during PAL APIs. The system call errors may include
```

Comment:
Is this going to work well with thread suspension? You may get number of extra PAL API calls if a thread suspension happens during the PInvoke transitions. It may be confusing to see these unrelated calls as part of the error message.

Reply:
Good point, maybe the tracking can be paused/resumed around suspension, I'll update.

Reply:
I think it may be better to do the tracking for the mutex implementation only (pass the error tracing context explicitly instead of implicitly as thread static).

Reply:
Currently only APIs relevant to named mutexes track errors. I thought it could be expanded to include other APIs if needed. Passing in a tracing context would work too, but what would be the advantage of doing that (wouldn't it end up getting stored on the thread anyway)?

Reply:
An alternative I considered is to use separate APIs for the mutex-specific ones that include something like a tracing context, but that wouldn't cover waits and it wouldn't be easily expandable. I felt this solution would be simpler and reusable.

Reply:
We have been working on making the Win32 PAL to disappear. The machine-wide mutex is one of the few remaining substantial pieces remaining in the Win32 PAL. Let's consider what the picture may look like if it was moved to some combination of System.Native and C#. It is unlikely we would want to build System.Native-wide syscall tracing mechanism, there is no need for it and it would just add overhead. It would likely be error reporting specific for the machine-wide mutex operations.

> wouldn't it end up getting stored on the thread anyway

In general, it is better to pass things around explicitly instead of implicitly via thread statics when there is a choice. If you have a strong preference for thread statics here, I can live with it.

Reply:
Ok how about the following for now?
- Create separate qcalls for creating/opening/releasing mutexes on Unix
- Refactor Mutex.Windows.cs into common/Windows/Unix portions
- Pass in something to the Unix APIs to track system call errors

Reply:
Sounds good to me.

> Mutex.Windows.cs into common/Windows/Unix portions

Nit: Mutex.Windows.cs is <100 lines and the common part is going to be a small fraction of it. To keep things simple, I would just create `Mutex.CoreCLR.Windows.cs` under src\coreclr\system.private.corelib and leave Mutex.Windows.cs alone (maybe just delete `TARGET_UNIX` ifdefs from it since they are not going to be needed anymore).

Reply:
> Nit: Mutex.Windows.cs is <100 lines and the common part is going to be a small fraction of it. To keep things simple, I would just create `Mutex.CoreCLR.Windows.cs` under src\coreclr\system.private.corelib and leave Mutex.Windows.cs alone (maybe just delete `TARGET_UNIX` ifdefs from it since they are not going to be needed anymore).

It seems like we can just fork the API call and keep the common portion mostly the same between runtimes on Windows, I've copied the code for now in the latest commit.